packages/nitro-protocol/docs/force-move-safety.md

```diff
@@ -0,0 +1,399 @@
+If his transaction is recorded first, it might change the state in such a way that Alice's transaction reverts.
+
+# Gas-optimization
+
+The situation is worsened by an optimization we've done to save on gas fees.
+A hash of the entire channel state is stored in a single `bytes32` slot on-chain.
+Therefore, when applying an action, the participant must provide the current state as a part of the calldata.
+If the hash of the provide state does not match the stored value, the action has no effect.
+
+| State       | Action             | NextState   | Requirements |
+| ----------- | ------------------ | ----------- | ------------ |
+| Open(n)     | forceMove(m,s,p,S) | Chal(m,s,p) | matches(S)   |
+|             |                    |             | m >= n       |
+| Chal(n,s,p) | refute(m,s',S)     | Open(n)     | matches(S)   |
+|             |                    |             | m > n        |
+|             |                    |             | p == sig(s') |
+| Chal(n,s,p) | respond(s',S)      | Open(n+1)   | matches(S)   |
+|             |                    |             | s->s'        |
+| Open(n)     | respondAlt(S)      | Open(n+1)   | matches(S)   |
+
+```
+╔═════════════╦════════════════════╦═════════════╦══════════════╗
+║ State       ║ Action             ║ NextState   ║ Requirements ║
+╠═════════════╬════════════════════╬═════════════╬══════════════╣
+║ Open(n)     ║ forceMove(m,s,p,S) ║ Chal(m,s,p) ║ matches(S)   ║
+║             ║                    ║             ╠══════════════╣
+║             ║                    ║             ║ m >= n       ║
+╠═════════════╬════════════════════╬═════════════╬══════════════╣
+║ Chal(n,s,p) ║ refute(m,s',S)     ║ Open(n)     ║ matches(S)   ║
+║             ║                    ║             ╠══════════════╣
+║             ║                    ║             ║ m > n        ║
+║             ║                    ║             ╠══════════════╣
+║             ║                    ║             ║ p == sig(s') ║
+╠═════════════╬════════════════════╬═════════════╬══════════════╣
+║ Chal(n,s,p) ║ respond(s',S)      ║ Open(n+1)   ║ matches(S)   ║
+║             ║                    ║             ╠══════════════╣
+║             ║                    ║             ║ s->s'        ║
+╠═════════════╬════════════════════╬═════════════╬══════════════╣
+║ Open(n)     ║ respondAlt(S)      ║ Open(n+1)   ║ matches(S)   ║
+╚═════════════╩════════════════════╩═════════════╩══════════════╝
+```
+
+```
+function matches(S) {
+  return keccack256(S) == currentState;
+}
+```
+
+## Attack
+
+- Suppose the channel is in `Open(k)`
+- Alice wishes to apply `forceMove(n,s,Alice)`
+- Bob front-runs with `forceMove(k,s',Eve)`
+
+```
+   forceMove(n,s,Alice,S) >> forceMove(k,s',Eve,S) >> Open(k)
+== forceMove(n,s,Alice,S) >> Challenge(k,s',Eve)
+== Challenge(k,s',Eve)
+```
+
```

Comment:
I think this notation is great; but it might be nice to remind the reader here and explicitly mention that `A>>B>>S` denotes tx `A` being front-run by tx `B`.

Reply:
agree - that took me a while to work out